discovery/lib/controller-runtime/source/dynamic.go

```diff
@@ -0,0 +1,177 @@
+package source
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	ctrlcache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("source")
+
+type Dynamic struct {
+	// Injected by the controller manager
+	client discovery.ServerResourcesInterface
+	cache  ctrlcache.Cache
+	stop   <-chan struct{}
+
+	mu     sync.RWMutex
+	active map[schema.GroupVersionKind]source.Source
+}
+
+var (
+	_ source.Source    = &Dynamic{}
+	_ inject.Cache     = &Dynamic{}
+	_ inject.Stoppable = &Dynamic{}
+)
+
+// Start is internal and should be called only by the Controller to register an EventHandler with the Informer to enqueue reconcile.Requests.
+func (d *Dynamic) Start(handle handler.EventHandler, queue workqueue.RateLimitingInterface, predicates ...predicate.Predicate) error {
+	if d.client == nil {
+		return fmt.Errorf("must inject rest config before starting dynamic source")
+	}
+	if d.cache == nil {
+		return fmt.Errorf("must inject cache before starting dynamic source")
+	}
+	if d.stop == nil {
+		return fmt.Errorf("must inject stop channel before starting dynamic source")
+	}
+
+	// TODO(njhale): Build an aggregate watch implementation that tracks all APIResourceList for available groups.
+	// TODO(njhale): Stop sources when GVKs are removed
```

Comment:
This falls down when GVKs are removed from API discovery (e.g. a CRD is deleted). I would like this to play nice with controller-runtime abstractions, but that might not be possible. Any ideas?

Reply:
@njhale Is the idea that the Informer is specific to a single GVK?
Do you have ideas on how stopping it could be addressed?

Reply:
@stealthybox In abstract, the idea is to provide a single event source that emits resources for all GVKs available in API discovery (for each point in time). There's also an additional requirement that this event source populates controller-runtime's client cache. Effectively, this implementation fans-in multiple informers to achieve this.

> Do you have ideas on how stopping it could be addressed?

There's a [controller-runtime issue](https://github.com/kubernetes-sigs/controller-runtime/issues/935) and a [PR](https://github.com/kubernetes-sigs/controller-runtime/pull/936) that addresses this directly, though it seems like it's quickly becoming a broader conversation involving apimachinery.